File: src/java/magma/app/compile/type/resolve/ResolvedTypes.java
package magma.app.compile.type.resolve;

import magma.api.collect.list.List;
import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;
import magma.app.Location;
import magma.app.Platform;
import magma.app.compile.CompileState;
import magma.app.compile.Dependency;
import magma.app.compile.Import;
import magma.app.compile.Registry;
import magma.app.io.Source;

public final class ResolvedTypes {
    public static CompileState getState(CompileState state, Location location) {
        var requestedNamespace = location.namespace();
        var requestedChild = location.name();

        var namespace = ResolvedTypes.fixNamespace(requestedNamespace, state.context().findNamespaceOrEmpty());
        if (state.registry().doesImportExistAlready(requestedChild)) {
            return state;
        }

        var namespaceWithChild = namespace.addLast(requestedChild);
        var anImport = new Import(namespaceWithChild, requestedChild);
        return state.mapRegistry((Registry registry) -> registry.addImport(anImport));
    }

    public static CompileState addResolvedImportFromCache0(CompileState state, String base) {
        if (state.stack().hasAnyStructureName(base)) {
            return state;
        }

        return state.context()
                .findSource(base)
                .map((Source source) -> {
                    Location location = source.createLocation();
                    return ResolvedTypes.getCompileState1(state, location)
                            .orElseGet(() -> ResolvedTypes.getState(state, location));
                })
                .orElse(state);
    }

    private static Option<CompileState> getCompileState1(CompileState immutableCompileState, Location location) {
        if (!immutableCompileState.context().hasPlatform(Platform.PlantUML)) {
            return new None<CompileState>();
        }

        var name = immutableCompileState.context().findNameOrEmpty();
        var dependency = new Dependency(name, location.name());
        if (immutableCompileState.registry().containsDependency(dependency)) {
            return new None<CompileState>();
        }

        return new Some<CompileState>(immutableCompileState.mapRegistry((Registry registry1) -> registry1.addDependency(dependency)));
    }

    private static List<String> fixNamespace(List<String> requestedNamespace, List<String> thisNamespace) {
        if (thisNamespace.isEmpty()) {
            return requestedNamespace.addFirst(".");
        }

        return ResolvedTypes.addParentSeparator(requestedNamespace, thisNamespace.size());
    }

    private static List<String> addParentSeparator(List<String> newNamespace, int count) {
        var index = 0;
        var copy = newNamespace;
        while (index < count) {
            copy = copy.addFirst("..");
            index++;
        }

        return copy;
    }
}

File: src/ts/magma/app/compile/type/resolve/ResolvedTypes.ts
import { CompileState } from "../../../../../magma/app/compile/CompileState";
import { Location } from "../../../../../magma/app/Location";
import { Import } from "../../../../../magma/app/compile/Import";
import { Registry } from "../../../../../magma/app/compile/Registry";
import { Source } from "../../../../../magma/app/io/Source";
import { Option } from "../../../../../magma/api/option/Option";
import { Platform } from "../../../../../magma/app/Platform";
import { None } from "../../../../../magma/api/option/None";
import { Dependency } from "../../../../../magma/app/compile/Dependency";
import { Some } from "../../../../../magma/api/option/Some";
import { List } from "../../../../../magma/api/collect/list/List";
export class ResolvedTypes {
	static getState(state: CompileState, location: Location): CompileState {
		let requestedNamespace = location.namespace()/*unknown*/;
		let requestedChild = location.name()/*unknown*/;
		let namespace = fixNamespace(requestedNamespace, state.context().findNamespaceOrEmpty())/*unknown*/;
		if (state.registry().doesImportExistAlready(requestedChild)/*unknown*/){
			return state/*CompileState*/;
		}
		let namespaceWithChild = namespace.addLast(requestedChild)/*unknown*/;
		let anImport = new Import(namespaceWithChild, requestedChild)/*unknown*/;
		return state.mapRegistry((registry: Registry) => registry.addImport(anImport)/*unknown*/)/*unknown*/;
	}
	static addResolvedImportFromCache0(state: CompileState, base: string): CompileState {
		if (state.stack().hasAnyStructureName(base)/*unknown*/){
			return state/*CompileState*/;
		}
		return state.context().findSource(base).map((source: Source) => {
			let location: Location = source.createLocation()/*unknown*/;
			return getCompileState1(state, location).orElseGet(() => getState(state, location)/*unknown*/)/*unknown*/;
		}).orElse(state)/*unknown*/;
	}
	static getCompileState1(immutableCompileState: CompileState, location: Location): Option<CompileState> {
		if (!!immutableCompileState/*CompileState*/.context().hasPlatform(Platform.PlantUML)/*unknown*/){
			return new None<CompileState>()/*unknown*/;
		}
		let name = immutableCompileState.context().findNameOrEmpty()/*unknown*/;
		let dependency = new Dependency(name, location.name())/*unknown*/;
		if (immutableCompileState.registry().containsDependency(dependency)/*unknown*/){
			return new None<CompileState>()/*unknown*/;
		}
		return new Some<CompileState>(immutableCompileState.mapRegistry((registry1: Registry) => registry1.addDependency(dependency)/*unknown*/))/*unknown*/;
	}
	static fixNamespace(requestedNamespace: List<string>, thisNamespace: List<string>): List<string> {
		if (thisNamespace.isEmpty()/*unknown*/){
			return requestedNamespace.addFirst(".")/*unknown*/;
		}
		return addParentSeparator(requestedNamespace, thisNamespace.size())/*unknown*/;
	}
	static addParentSeparator(newNamespace: List<string>, count: number): List<string> {
		let index = 0/*unknown*/;
		let copy = newNamespace/*List<string>*/;
		while (index < count/*unknown*/){
			copy/*unknown*/ = copy.addFirst("..")/*unknown*/;
			index/*unknown*/++;
		}
		return copy/*unknown*/;
	}
}
